zerver/lib/outgoing_webhook.py

```diff
@@ -8,17 +8,46 @@
 import logging
 from six.moves import urllib
 from functools import reduce
+from requests import Response
 
 from django.utils.translation import ugettext as _
 
-from zerver.models import Realm, get_realm_by_email_domain, get_user_profile_by_id, get_client
+from zerver.models import Realm, UserProfile, get_realm_by_email_domain, get_user_profile_by_id, get_client
 from zerver.lib.actions import check_send_message
 from zerver.lib.queue import queue_json_publish
 from zerver.lib.validator import check_dict, check_string
 from zerver.decorator import JsonableError
 
 MAX_REQUEST_RETRIES = 3
 
+class OutgoingWebhookServiceInterface(object):
+
+    def __init__(self, base_url, token, user_profile, service_name):
+        # type: (Text, Text, UserProfile, Text) -> None
+        self.base_url = base_url  # type: Text
+        self.token = token  # type: Text
+        self.user_profile = user_profile  # type: Text
+        self.service_name = service_name  # type: Text
+
+    # would receive the data generated on triggering of outgoing webhooks like service name, trigger method, message
+    # object, etc and return rest operation related info like relative url, base url, method of query i.e. POST, GET,
+    # etc and data to be sent to third party url.
+    def process_command(self, event):
+        # type: (Dict[str, Any]) -> Tuple[Dict[str ,Any], Dict[str, Any]]
+        raise NotImplementedError()
+
+    # in case of success of request, it would parse the response obtained and return the content of message
```

Comment:
This comment is already very helpful, but I think it's possible to make it even more immediately clear.  Here's what I might write:

```
# Given a successful response to the outgoing webhook REST operation, returns the message
# that should be sent back to the user.
#
# The response will have the following format: ...
# The event will be the same as the input to process_command.
# The returned message should have the following format: ...
```

Again, the goal is for the user (either someone who wants to use this interface, or someone who wants to write a new implementation of the interface) to have all the information they need in one place.